### PR TITLE
#3072 - e-Cert errors report - Part 1

### PR DESCRIPTION
**Pull request by guru-aot**
ECert feedback error report for ministry user in the report page.

 - date_received from sims.disbursement_feedback_errors is used to start_date and end_date filters in the report generation
 - Offering intensity of both parttime and fulltime is selected only one report is generated, as previous reports.
Sample report looks like below
- No block_funding check is done in the query so ALL error codes are exported  as per the acceptance criteria.

image

E2E test cases will be done in the second PR.

### CODE REVIEW

File: sources/packages/backend/apps/db-migrations/src/migrations/1717081412569-CreateECertErrorsReport.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+{
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      getSQLFileData("Create-eCert-errors-report.sql", "Reports"),
+    );
+  }
+
+  public async down(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      getSQLFileData("Drop-eCert-errors-report.sql", "Reports"),
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/db-migrations/src/migrations/1717081412569-CreateECertErrorsReport.ts:15]: Please adjust the rollback file name as `Rollback-create-e-cert-errors-report.sql`

File: sources/packages/backend/apps/db-migrations/src/migrations/1717081412569-CreateECertErrorsReport.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { MigrationInterface, QueryRunner } from "typeorm";
+import { getSQLFileData } from "../utilities/sqlLoader";
+
+export class CreateECertErrorsReport1717081412569
+  implements MigrationInterface
+{
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      getSQLFileData("Create-eCert-errors-report.sql", "Reports"),
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/db-migrations/src/migrations/1717081412569-CreateECertErrorsReport.ts:9]: Please adjust the name as `Create-e-cert-errors-report.sql`

File: sources/packages/backend/apps/db-migrations/src/sql/Reports/Create-eCert-errors-report.sql [hunk 3]
```diff
@@ -0,0 +1,49 @@
+INSERT INTO
+    sims.report_configs (report_name, report_sql)
+VALUES
+    (
+        'ECert_Errors_Report',
+        'select
+            disbursement_schedules.id as "eCert Number",
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Reports/Create-eCert-errors-report.sql:7]: I believe the `eCert Number` should be the e-Cert document number, please check.

File: sources/packages/backend/apps/db-migrations/src/sql/Reports/Create-eCert-errors-report.sql
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Reports/Create-eCert-errors-report.sql:1]: Is the SQL formatted by DBeaver of by VSCode?
Formatting in VSCode looks like the below for me.
```sql
SELECT
    disbursement_schedules.id AS "eCert Number",
    applications.application_number AS "Application Number",
    users.first_name AS "First Name",
    users.last_name AS "Last Name",
    students.birth_date AS "Date of Birth",
    disbursement_feedback_errors.feedback_file_name AS "Feedback File Name",
    education_programs_offerings.study_start_date AS "Study Start Date",
    education_programs_offerings.study_end_date AS "Study End Date",
    to_char(
        disbursement_feedback_errors.date_received,
        'YYYY-MM-DD'
    ) AS "Error Logged Date",
    STRING_AGG(ecert_feedback_errors.error_code, ', ') AS "Error Codes"
FROM
    sims.disbursement_feedback_errors
    INNER JOIN sims.ecert_feedback_errors ON ecert_feedback_errors.id = disbursement_feedback_errors.ecert_feedback_error_id
    INNER JOIN sims.disbursement_schedules ON disbursement_schedules.id = disbursement_feedback_errors.disbursement_schedule_id
    INNER JOIN sims.student_assessments ON student_assessments.id = disbursement_schedules.student_assessment_id
    INNER JOIN sims.education_programs_offerings ON education_programs_offerings.id = student_assessments.offering_id
    INNER JOIN sims.applications ON applications.id = student_assessments.application_id
    INNER JOIN sims.students ON students.id = applications.student_id
    INNER JOIN sims.users ON users.id = students.user_id
WHERE
    education_programs_offerings.offering_intensity = any(:offeringIntensity)
    AND disbursement_feedback_errors.date_received BETWEEN :startDate
    AND :endDate
GROUP BY
    disbursement_schedules.id,
    applications.application_number,
    users.first_name,
    users.last_name,
    students.birth_date,
    disbursement_feedback_errors.feedback_file_name,
    education_programs_offerings.study_start_date,
    education_programs_offerings.study_end_date,
    disbursement_feedback_errors.date_received
ORDER BY
    disbursement_schedules.id;
```

File: sources/packages/backend/apps/db-migrations/src/sql/Reports/Create-eCert-errors-report.sql [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            applications.application_number as "Application Number",
+            users.first_name as "First Name",
+            users.last_name as "Last Name",
+            students.birth_date as "Date of Birth",
+            disbursement_feedback_errors.feedback_file_name as "Feedback File Name",
+            education_programs_offerings.study_start_date as "Study Start Date",
+            education_programs_offerings.study_end_date as "Study End Date",
+            to_char(disbursement_feedback_errors.date_received, ''YYYY-MM-DD'') as "Error Logged Date",
+            STRING_AGG(ecert_feedback_errors.error_code,'', '') as "Error Codes"
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Reports/Create-eCert-errors-report.sql:16]: Pretty cool 👍

**Review by andrewsignori-aot — COMMENTED** [review]
Great work, please take a look at the comments.

File: sources/packages/backend/apps/db-migrations/src/sql/Reports/Create-e-cert-errors-report.sql [hunk 5]
```diff
@@ -0,0 +1,45 @@
+INSERT INTO
+    sims.report_configs (report_name, report_sql)
+VALUES
+    (
+        'ECert_Errors_Report',
+        'select
+            disbursement_schedules.document_number as "eCert Number",
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Reports/Create-e-cert-errors-report.sql:7]: Minor, the AC states to be named as "eCert #".

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for making the changes, looks good 👍 
Please ensure the column names are reflecting the ACs or the business is ok with the differences.

**Comment by guru-aot** [comment]
> Thanks for making the changes, looks good 👍 Please ensure the column names are reflecting the ACs or the business is ok with the differences.

I just verified the generated excel with @CarlyCotton.

**Pull request merged**